shared/src/main/scala/com/raquo/waypoint/Route.scala

```diff
@@ -365,36 +454,45 @@ object Route {
     decode: PartialFunction[FragmentPatternArgs[PathArgs, QueryArgs, FragmentArgs], Page],
     pattern: PathQueryFragmentRepr[PathArgs, DummyError, QueryArgs, DummyError, FragmentArgs, DummyError],
     basePath: String = ""
-  ): Route[Page, FragmentPatternArgs[PathArgs, QueryArgs, FragmentArgs]] = {
-    new Route(
+  ): Partial[Page, FragmentPatternArgs[PathArgs, QueryArgs, FragmentArgs]] = {
+    withQueryAndFragmentRouteBuilder(pattern)(b => new Partial(
       matchEncodePF = matchEncode,
       decodePF = decode,
-      createRelativeUrl = { args =>
-        val patternArgs = PathQueryFragmentMatching(path = args.path, query = args.query, fragment = args.fragment)
-        "/" + pattern.createPart(patternArgs)
-      },
-      matchRelativeUrl = relativeUrl => pattern.matchRawUrl(relativeUrl).toOption,
+      createRelativeUrl = b.createRelativeUrl,
+      matchRelativeUrl = b.matchRelativeUrl,
       basePath = basePath
-    )
+    ))
```

Comment:
So, previously, `withQueryAndFragment` would call into `withQueryAndFragmentPF`, but now that's not possible because we have specialized Partial and Total classes for the two cases.

In this PR, to keep sharing the code between those implementations, you've introduced the `RouteBuilder` abstraction. After looking at it, I don't think it is carrying its weight. With only two usages of `withQueryAndFragmentRouteBuilder`, we are actually spending more lines of code (incl. the definition of `withQueryAndFragmentRouteBuilder`) compared to just not sharing the implementations of `withQueryAndFragment` and `withQueryAndFragmentPF`. For me, both shorter code, and fewer abstractions / indirections is better, even if it's a bit repetitive.

The same goes for other uses of `RouteBuilder`. IIUC, they all exist only for code sharing, but since they produce more lines of code, and more abstracted / indirect code, I think the `RouteBuilder` abstraction should be eliminated. Implementations should just spell out everything with no regard for code sharing, like the old `def withQueryAndFragmentPF`, `def onlyQueryAndFragmentPF`, etc. do.

Reply:
I originally did what you said, but then because all of the things are private inside Route, you have to have `private[waypoint] def toTotal: Route.Total[Page, Args]` in there and then you have to explicitly provide type arguments in the `Total` constructors because Scala couldn't infer them from type signature anymore, which felt just as repetive.

While the builder currently doesn't do much IMHO it's fundamentally sensible, especially if more shared code is added in the future. The other approach is just copy-pasting code, which is easy but allows for errors when modifying code. I suppose you can argue that tests should catch it?

Reply:
I'm not sure what you mean with `def toTotal` – would that be a method on Partial? Why would we need it?

What I meant, for example, is that `def onlyQuery` could be just something like:

```scala
  new Total(
      encode = encode,
      decode = decode,
      createRelativeUrl = args => "/" + pattern.createUrlString(path = (), params = args),
      matchRelativeUrl = relativeUrl => pattern.matchRawUrl(relativeUrl).toOption.map(_.params),
      basePath = basePath
  ))
```

and `def onlyQueryPF` could have a very similar but non-shared implementation with `new Partial(...)`. It would repeat the createRelativeUrl / matchRelativeUrl code, but those are just two usages of two lines, I'm fine with that to reduce abstraction and make the code shorter.